src/coreclr/src/jit/liveness.cpp

```diff
@@ -349,11 +349,10 @@ void Compiler::fgPerNodeLocalVarLiveness(GenTree* tree)
             // This ensures that the block->bbVarUse will contain
             // the FrameRoot local var if is it a tracked variable.
 
-            if ((tree->AsCall()->IsUnmanaged() || tree->AsCall()->IsTailCallViaJitHelper()) &&
-                compMethodRequiresPInvokeFrame())
+            if ((call->IsUnmanaged() || call->IsTailCallViaJitHelper()) && compMethodRequiresPInvokeFrame())
             {
                 assert((!opts.ShouldUsePInvokeHelpers()) || (info.compLvFrameListRoot == BAD_VAR_NUM));
-                if (!opts.ShouldUsePInvokeHelpers())
+                if (!opts.ShouldUsePInvokeHelpers() && !call->IsSuppressGCTransition())
```

Comment:
Could this predicate be rewritten as:

```c++
if ((info.compLvFrameListRoot != BAD_VAR_NUM) && !call->IsSuppressGCTransition())
```

Since the member `compLvFrameListRoot` is directly used it seems clearer the relationship between the predicate and the branch logic.

Reply:
It looks like the assumption here is very much `info.compLvFrameListRoot != BAD_VAR_NUM` - in subsequent areas as well. If `ShouldUsePInvokeHelpers` is false does this mean that we should always have a FrameListRoot?

Reply:
> Since the member compLvFrameListRoot is directly used it seems clearer the relationship between the predicate and the branch logic.

Actually, I think the logic is clearer that `compLvFrameListRoot` should be marked as used if the logic indicates it is needed (i.e. that `ShouldUsePInvokeHelpers` is true). In any event, the relationship between the two should be quite clear from the assert above.

> If ShouldUsePInvokeHelpers is false does this mean that we should always have a FrameListRoot?

Only if there are unmanaged calls that require it. (See `compMethodRequiresPInvokeFrame()` which is defined as `return (info.compUnmanagedCallCountWithGCTransition > 0);`